packages/cli-repl/src/cli-repl.ts

```diff
@@ -295,56 +171,33 @@ class CliRepl {
    * @param {string} driverUrl - The driver URI.
    * @param {NodeOptions} driverOptions - The driver options.
    */
-  requirePassword(driverUri: string, driverOptions: NodeOptions): void {
+  async requirePassword(driverUri: string, driverOptions: NodeOptions): Promise<void> {
     const readOptions = {
       prompt: 'Enter password: ',
       silent: true,
-      replace: '*'
+      replace: '*',
+      input: this.input,
+      output: this.output
     };
-    read(readOptions, (error, password) => {
-      if (error) {
-        this.bus.emit('mongosh:error', error);
-        return console.log(this.formatError(error));
-      }
-
-      (driverOptions.auth as any).password = password;
-      this.setupRepl(driverUri, driverOptions).catch((e) => {
-        this._fatalError(e);
-      });
-    });
-  }
-
-  private _fatalError(error: any): never {
-    if (this.bus) {
+    try {
+      (driverOptions.auth as any).password = await promisify(read)(readOptions);
+    } catch (error) {
       this.bus.emit('mongosh:error', error);
+      return this._fatalError(error);
     }
-
-    console.error(this.formatError(error));
-    return process.exit(1);
   }
 
-  onPrint(values: ShellResult[]): void {
-    const joined = values.map(this.writer).join(' ');
-    this.repl.output.write(joined + '\n');
-  }
-
-  formatOutput(value: any): string {
-    return formatOutput(value, this.getFormatOptions());
-  }
+  private _fatalError(error: any): never {
+    this.bus.emit('mongosh:error', error);
 
-  formatError(value: any): string {
-    return formatError(value, this.getFormatOptions());
+    this.output.write(this.mongoshRepl.formatError(error) + '\n');
+    this.bus.emit('mongosh:exit', 1);
```

Comment:
Is there a reason we are doing this with the bus and not in some more explicit way, like:

``` js
this.exit(1);
```

Maybe even injecting `process` if that was the issue. Even if not I'd wrap `this.bus.emit('mongosh:exit')` in `exit(statusCode)` regardless.

Reply:
The reason that this uses the bus is that it keeps us from having to add yet another property to the `MongoshRepl` and that it’s a bit easier to test this way, but if we want to, we can do that, sure.



> Even if not I'd wrap `this.bus.emit('mongosh:exit')` in `exit(statusCode)` regardless.

Sure, done! :)